p2p/src/authenticated/actors/peer/actor.rs

```diff
@@ -69,17 +71,48 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Rng + CryptoRng> Actor<E> {
     ) -> Result<(), Error> {
         trace!(peer = hex(peer), "sending data",);
 
-        // Send data
         let channel = data.channel;
+        Self::send_message(
+            sender,
+            wire::message::Payload::Data(data),
+            metrics::Message::new_data(peer, channel),
+            sent_messages,
+        )
+        .await?;
+        Ok(())
+    }
+
+    // validate_prioritized_data returns Data after few checks.
+    fn validate_prioritized_data<T>(
+        msg: Option<wire::Data>,
+        contains: T,
+    ) -> Result<wire::Data, Error>
+    where
+        T: Fn(&u32) -> bool,
+    {
+        let data = match msg {
+            Some(data) => data,
+            None => return Err(Error::PeerDisconnected),
+        };
+        if !contains(&data.channel) {
+            return Err(Error::InvalidChannel);
+        }
+        Ok(data)
+    }
+
+    // send_message creates a message from a payload, then sends and increments metrics.
+    async fn send_message<Si: Sink>(
+        sender: &mut Sender<Si>,
+        payload: Payload,
+        metrics_message: metrics::Message,
+        sent_messages: &Family<metrics::Message, Counter>,
+    ) -> Result<(), Error> {
         let msg = wire::Message {
-            payload: Some(wire::message::Payload::Data(data)),
+            payload: Some(payload),
         }
         .encode_to_vec();
         sender.send(&msg).await.map_err(Error::SendFailed)?;
-        sent_messages
-            .get_or_create(&metrics::Message::new_data(peer, channel))
-            .inc();
-
+        sent_messages.get_or_create(&metrics_message).inc();
```

Comment:
I wanted to use `this.sent_messages` instead of passing as function arg, but it would require the function to have a `&self` as first arg, then `async move` raises an alert. Not an expert of using `async move`, I keep like this right now, let me know if there is a clever way to access the variable.